### PR TITLE
[FIRRTL] Add FlowKind op interface

### PR DESCRIPTION
**Pull request by youngar**
This is an op interface to calculate the flow of operations. When an
operation has flow other than `Source`, it can implement this interface
to customize its flow kind.

This is helpful because it makes flow checking of the FIRRTL dialect
extensible.  This is needed to be able to move the CHIRRTL operations
into their own dialect without having the FIRRTL dialect depend on the
CHIRRTL dialect, creating a circular dependency.

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td [hunk 1]
```diff
@@ -12,6 +12,59 @@
 
 include "mlir/IR/OpBase.td"
 
+def FlowKind : OpInterface<"FlowKind"> {
+  let cppNamespace = "circt::firrtl";
+  let description = [{
+    Allows operations to specify what kind of flow they have.  When an operation
+    does not implement this interface, they are assumed to have "source" flow.
+
+    If an operation has results, they should implement `getFlow`.  If they
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:21]: Missing reset of the description?

File: lib/Dialect/FIRRTL/FIRRTLOpInterfaces.cpp [hunk 2]
```diff
@@ -20,6 +20,63 @@ using namespace mlir;
...
+    } else {
+      auto op = dyn_cast<FlowKind>(value.getDefiningOp());
+      // If the operation doesn't implement the interface, return the result
+      // so far.
+      if (!op)
+        return flow;
+      auto result = op.getFlow(value.cast<OpResult>());
+      flow = mergeFlow(result.getFlow(), flow);
+      value = result.getValue();
+    }
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLOpInterfaces.cpp:71]: You may be able to share some code here by defining a `FlowResult result;` outside the if/else, setting that inside, and then keeping `flow = ...;' and `value = ...;' common after the if/else.

**Review by seldridge — APPROVED** [review]
This looks super clean to me. Nice work!

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td (hunk 1 above)
Comment (youngar) [reply include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td]: Woops, I meant to delete that part.. I had a more complicated interface originally that I thought required some more detailed instructions

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 3]
```diff
@@ -1671,7 +1628,7 @@ static LogicalResult verifyPartialConnectOp(PartialConnectOp partialConnect) {
            << ". Types are not weakly equivalent.";
 
   // Check that the flows make sense.
-  if (foldFlow(partialConnect.src()) == Flow::Sink) {
+  if (FlowKind::get(partialConnect.src()) == Flow::Sink) {
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1631]: possible simplification: FlowKind::isSink(val)?

**Review by darthscsi — APPROVED** [review]
LGTM

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 3 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1631]: I like it, added for all three kinds of flow.

**Comment by lattner** [comment]
Random thought: while the chirrtl and firrtl dialects are theoretically separated, there is nothing that actually benefits us having one without the other.  Would it make sense to keep the two dialects distinct, but put them in the same library?  This would allow you to have circular references etc.

We have a similar issue going on in the hw dialect, where we put some SV stuff into it to make the implementation simpler.

**Comment by youngar** [comment]
This is no longer needed.

**Pull request closed without merging**